File: backend/src/models/Pet.js
const mongoose = require('mongoose');

const petSchema = new mongoose.Schema({
  name: {
    type: String,
    required: [true, '请输入宠物名字'],
    trim: true,
    maxlength: [50, '宠物名字不能超过50个字符']
  },
  breed: {
    type: String,
    required: [true, '请选择宠物品种'],
    trim: true
  },
  birthday: {
    type: Date,
    required: [true, '请选择宠物生日']
  },
  gender: {
    type: String,
    enum: ['male', 'female', 'unknown'],
    default: 'unknown'
  },
  avatar: {
    type: String,
    default: '' // 头像URL，可选
  },
  owner: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'User',
    required: true
  },
  bio: {
    type: String,
    maxlength: [500, '个人简介不能超过500个字符'],
    default: ''
  },
  weight: {
    type: Number, // 体重（kg）
    min: 0
  },
  isActive: {
    type: Boolean,
    default: true
  }
}, {
  timestamps: true, // 自动添加createdAt和updatedAt
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// 虚拟字段：计算宠物年龄
petSchema.virtual('age').get(function() {
  if (!this.birthday) return 0;
  const today = new Date();
  const birthDate = new Date(this.birthday);
  let age = today.getFullYear() - birthDate.getFullYear();
  const monthDiff = today.getMonth() - birthDate.getMonth();
  if (monthDiff < 0 || (monthDiff === 0 && today.getDate() < birthDate.getDate())) {
    age--;
  }
  return age;
});

// 索引优化
petSchema.index({ owner: 1, isActive: 1 });
petSchema.index({ createdAt: -1 });

module.exports = mongoose.model('Pet', petSchema);

File: dogreal_app/lib/models/pet.dart
class Pet {
  final String id;
  final String name;
  final String breed;
  final DateTime birthday;
  final String gender;
  final String avatar;
  final String owner;
  final String bio;
  final double? weight;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;
  final int age;

  Pet({
    required this.id,
    required this.name,
    required this.breed,
    required this.birthday,
    required this.gender,
    required this.avatar,
    required this.owner,
    required this.bio,
    this.weight,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
    required this.age,
  });

  factory Pet.fromJson(Map<String, dynamic> json) {
    return Pet(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      breed: json['breed'] ?? '',
      birthday: DateTime.parse(json['birthday']),
      gender: json['gender'] ?? 'unknown',
      avatar: json['avatar'] ?? '',
      owner: json['owner'] is String 
          ? json['owner'] 
          : (json['owner']?['_id'] ?? ''),
      bio: json['bio'] ?? '',
      weight: json['weight']?.toDouble(),
      isActive: json['isActive'] ?? true,
      createdAt: DateTime.parse(json['createdAt']),
      updatedAt: DateTime.parse(json['updatedAt']),
      age: json['age'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'name': name,
      'breed': breed,
      'birthday': birthday.toIso8601String(),
      'gender': gender,
      'avatar': avatar,
      'bio': bio,
      'weight': weight,
    };
  }

  Pet copyWith({
    String? id,
    String? name,
    String? breed,
    DateTime? birthday,
    String? gender,
    String? avatar,
    String? owner,
    String? bio,
    double? weight,
    bool? isActive,
    DateTime? createdAt,
    DateTime? updatedAt,
    int? age,
  }) {
    return Pet(
      id: id ?? this.id,
      name: name ?? this.name,
      breed: breed ?? this.breed,
      birthday: birthday ?? this.birthday,
      gender: gender ?? this.gender,
      avatar: avatar ?? this.avatar,
      owner: owner ?? this.owner,
      bio: bio ?? this.bio,
      weight: weight ?? this.weight,
      isActive: isActive ?? this.isActive,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      age: age ?? this.age,
    );
  }
}

class PetResponse {
  final bool success;
  final Pet? data;
  final String? message;

  PetResponse({
    required this.success,
    this.data,
    this.message,
  });

  factory PetResponse.fromJson(Map<String, dynamic> json) {
    return PetResponse(
      success: json['success'] ?? false,
      data: json['data'] != null ? Pet.fromJson(json['data']) : null,
      message: json['message'],
    );
  }
}
